thunder/distributed/transforms/fsdp_v2.py

```diff
@@ -153,92 +153,91 @@ def transform_module(
+                    device_adjustments[nn] = self.device
+
+        # Broadcast parameters if requested
+        if self.broadcast_from is not None:
+            if not is_fully_materialized:
+                # Note: we could move broadcasting into its own transform coming
+                #       after materialization (in thunder.distributed.fsdp) to
+                #       support this, if it is useful.
+                raise RuntimeError("cannot broadcast from non-materialized model")
+            for pn, p in chain(thunder_model.named_parameters(), thunder_model.named_buffers()):
+                tdist.broadcast(
+                    p,
+                    src=self.broadcast_from,
+                    group=self.process_group,
+                    async_op=False,
                 )
-                new_shape = thunder_model._overrides_parameters[pn].shape
-                self.sharded_params[pn] = (old_shape, new_shape, thunder_model._overrides_parameters[pn].device)
-                if self.release_original_parameters:
-                    base_pn = pn.rsplit(".", 1)[-1]
-                    p_orig = getattr(submodule, base_pn)
-                    if p_orig.device.type != "meta":
-                        p_meta = torch.nn.Parameter(p.to(device="meta"), requires_grad=p.requires_grad)
-                        p_meta._thunder_device = p_orig.device
-                        setattr(submodule, base_pn, p_meta)
-                    else:
-                        p_orig._thunder_device = self.device
-
-                # Track the original param and it's corresponding copied shard and metadata.
-                shared_params[p] = {
-                    "param_copy": thunder_model._overrides_parameters[pn],
-                    "param_shard_meta": self.sharded_params[pn],
-                    "param_name": pn,
-                }
+
+        # do the actual sharding. Note that meta tensors will give sharded meta tensors.
+        for pn, p in list(thunder_model.named_parameters()):
+            # we collect shapes and devices because we do not know if other transforms also change it.
+            old_shape = p.shape
+            p_new, _ = _shard_tensor(p, global_rank, world_size, pn, allow_padding_for_fsdp=True)
+            p_new = torch.nn.Parameter(p_new.clone(), requires_grad=p.requires_grad)
+            new_shape = p_new.shape
+            for nn in shared_names[pn]:
+                thunder_model._overrides_parameters[nn] = p_new
+                self.sharded_params[nn] = (old_shape, new_shape, getattr(p, "_thunder_device", p.device))
+            if self.release_original_parameters:
+                p_orig = thunder_model._model.get_parameter(pn)
+                if p_orig.device.type != "meta":
+                    p_meta = torch.nn.Parameter(p_orig.to(device="meta"), requires_grad=p_orig.requires_grad)
+                    p_meta._thunder_device = p_orig.device
+                    for nn in shared_names[pn]:
```

Comment:
what does "nn" stand for? IMHO nn is a common alias for `torch.nn` so I'd love to avoid using this abbreviation

Reply:
Good call. I'll change it after the rebase.

Reply:
fixed this.